go/worker/sentry/worker.go

```diff
@@ -0,0 +1,123 @@
+package sentry
+
+import (
+	"fmt"
+
+	flag "github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/oasislabs/oasis-core/go/common/grpc"
+	"github.com/oasislabs/oasis-core/go/common/identity"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	"github.com/oasislabs/oasis-core/go/sentry"
+	"github.com/oasislabs/oasis-core/go/sentry/api"
+)
+
+const (
+	CfgEnabled     = "worker.sentry.enabled"
+	CfgControlPort = "worker.sentry.control_port"
+)
+
+// Flags has the configuration flags.
+var Flags = flag.NewFlagSet("", flag.ContinueOnError)
+
+// Enabled returns true if Sentry worker is enabled.
+func Enabled() bool {
+	return viper.GetBool(CfgEnabled)
+}
+
+type Worker struct {
+	enabled bool
+
+	backend api.Backend
+
+	grpcServer *grpc.Server
+
+	quitCh chan struct{}
+	initCh chan struct{}
+
+	logger *logging.Logger
+}
+
+// Name returns the service name.
+func (w *Worker) Name() string {
+	return "sentry worker"
+}
+
+// Start starts the service.
+func (w *Worker) Start() error {
+	if !w.enabled {
+		w.logger.Info("not starting sentry worker as it is disabled")
+
+		return nil
+	}
+
+	// Start the sentry gRPC server.
+	if err := w.grpcServer.Start(); err != nil {
+		w.logger.Error("failed to start sentry gRPC server",
+			"err", err,
+		)
+		return err
+	}
+
+	close(w.initCh)
+
+	return nil
+}
+
+// Stop halts the service.
+func (w *Worker) Stop() {
+	if !w.enabled {
+		close(w.quitCh)
+		return
+	}
+
+	// TODO: How to close the gRPC server?
```

Comment:
Call `w.grpcServer.Stop()` and then during cleanup call `w.grpcServer.Cleanup()`.

Reply:
Thanks, done.